test/distributed/_spmd/test_data_parallel.py

```diff
@@ -0,0 +1,163 @@
+    def loss_fn(out, labels):
+        return (out - labels).sum()
+
+    optim.zero_grad()
+    inputs, labels = train_batch
+
+    out = model(inputs)
+    loss = loss_fn(out, labels)
+    loss.backward()
+    optim.step()
+    return loss
+
+
+class TestDataParallel(DTensorTestBase):
+    @property
+    def world_size(self):
+        return 2
+
+    def _test_data_parallel(
+        self, mod, ddp_mod, opt, ddp_opt, inp, train_step, data_parallel_mode
+    ):
+        ddp_inp = deepcopy(inp)
+
+        # need one step to warm up optimizers
+        train_step(mod, opt, inp)
+        opt.zero_grad()
+        # DDP run full train step once to align with the warmup
+        train_step(ddp_mod, ddp_opt, ddp_inp)
+        ddp_opt.zero_grad()
+
+        # train a DDP model once manually as DDP grads are different
+        torch.sum(ddp_mod(ddp_inp[0]) - ddp_inp[1]).backward()
+        with torch.no_grad():
+            for p in ddp_mod.parameters():
+                p.grad *= self.world_size
+        ddp_opt.step()
+
+        # compile it with replicate and run step once
+        compiled_fn = compile(parallel_mode=DataParallel(data_parallel_mode))(
+            train_step
+        )
+        compiled_fn(mod, opt, inp)
+
+        for p1, p2 in zip(mod.parameters(), ddp_mod.parameters()):
+            # mod parameters are DTensors, convert to local tensor before compare
+            if data_parallel_mode == "fully_shard":
+                # gather the shards for comparison
+                p1_replica = p1.redistribute(placements=[Replicate()])
+                p1_local_param = p1_replica.to_local()
+            else:
+                p1_local_param = p1.to_local()
+            self.assertEqual(p1_local_param, p2)
+
+    @skip_if_lt_x_gpu(2)
+    @with_comms
+    def test_replicate_sgd(self):
+        mod = SimpleMLP().cuda(self.rank)
+        opt = torch.optim.SGD(mod.parameters(), lr=0.1)
+        train_batch = (
+            torch.randn(128, 50).to(self.rank),
+            torch.randn(128, 8).to(self.rank),
+        )
+
+        ddp_mod = DDP(deepcopy(mod), device_ids=[self.rank])
+        ddp_opt = torch.optim.SGD(ddp_mod.parameters(), lr=0.1)
+
+        self._test_data_parallel(
+            mod, ddp_mod, opt, ddp_opt, train_batch, train_step, "replicate"
+        )
+
+    @skip_if_lt_x_gpu(2)
+    @with_comms
+    def test_replicate_sgd_momentum(self):
+        mod = SimpleMLP().cuda(self.rank)
+        opt = torch.optim.SGD(mod.parameters(), lr=0.1, momentum=0.9)
+
+        train_batch = (
+            torch.randn(128, 50).to(self.rank),
+            torch.randn(128, 8).to(self.rank),
+        )
+
+        ddp_mod = DDP(deepcopy(mod), device_ids=[self.rank])
+        ddp_opt = torch.optim.SGD(ddp_mod.parameters(), lr=0.1, momentum=0.9)
+
+        self._test_data_parallel(
+            mod, ddp_mod, opt, ddp_opt, train_batch, train_step, "replicate"
+        )
+
+    @skip_if_lt_x_gpu(2)
+    @with_comms
+    def test_fully_shard_sgd(self):
```

Comment:
nit: Maybe `test_fully_shard_sgd` and `test_fully_shard_sgd_momentum` can be refactored into a helper that takes in the optimizer kwargs to avoid so much duplication.

Reply:
Yeah I need to think a bit how to refactor the tests to avoid duplication, I'll create a follow up PR to clean it up.